applications/pelion_client/src/modules/factory_reset_request.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <zephyr.h>
+#include <event_manager.h>
+
+#include "init_factory_reset_event.h"
+
+#include <caf/events/button_event.h>
+
+#define MODULE factory_reset_request
+#include <caf/events/module_state_event.h>
+
+
+static struct k_delayed_work init_work;
+static bool factory_reset_requested;
+
+
+static void init_work_handler(struct k_work *work)
+{
+	struct init_factory_reset_event *event = new_init_factory_reset_event();
+
+	event->factory_reset = factory_reset_requested;
+	EVENT_SUBMIT(event);
+	module_set_state(MODULE_STATE_READY);
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_button_event(eh)) {
+		struct button_event *event = cast_button_event(eh);
+
+		if (k_delayed_work_pending(&init_work) &&  event->pressed) {
+			factory_reset_requested = true;
+		}
+		return false;
+	}
+
+	if (is_module_state_event(eh)) {
+		struct module_state_event *event = cast_module_state_event(eh);
+
+		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			k_delayed_work_init(&init_work, init_work_handler);
+			k_delayed_work_submit(&init_work, K_MSEC(50));
```

Comment:
We are not sure if button event would arrive in time.
For proper implementation, we could consider modifying CAF buttons to perform an initial scan and send button_events  before reporting MODULE_STATE_READY (in that way buttons pressed on init would always be properly recognized).
As a workaround we could consider making timeout defined as Kconfig option (e.g. if user adds another modules that react on main, he could expand this timeout).

Reply:
Timeout is now defined. Buttons module rewrite is in a plan for a future.